stan/math/laplace/laplace_marginal.hpp

```diff
@@ -0,0 +1,275 @@
+                            const K& covariance_function,
+                            double tolerance = 1e-6,
+                            long int max_num_steps = 100) {
+    Eigen::VectorXd theta, W_root, a, l_grad;
+    Eigen::MatrixXd L;
+    return laplace_marginal_density(theta_0, phi, x,
+              diff_likelihood, covariance_function,
+              theta, W_root, L, a, l_grad,
+              tolerance, max_num_steps);
+  }
+
+  /**
+   * A structure to compute sensitivities of the covariance
+   * function using forward mode autodiff.
+   */
+  template <typename K>
+  struct covariance_sensitivities {
+    std::vector<Eigen::VectorXd> x_;
+    int theta_size_;
+    K covariance_function_;
+
+    covariance_sensitivities (const std::vector<Eigen::VectorXd>& x,
+                              int theta_size,
+                              const K& covariance_function) :
+    x_(x), theta_size_(theta_size),
+    covariance_function_(covariance_function) { }
+
+    template <typename T>
+    Eigen::Matrix<T, Eigen::Dynamic, 1>
+    operator() (const Eigen::Matrix<T, Eigen::Dynamic, 1>& phi) const {
+      return to_vector(covariance_function_(phi, x_, theta_size_));
+    }
+  };
+
+  /**
+   * The vari class for the laplace marginal density.
+   * The method to compute gradients couples built-in forward autodiff
+   * with algorithm 5.1 in R & W.
+   * To make computation efficient, variables produced during the
+   * Newton step are stored and reused. 
+   */
+  struct laplace_marginal_density_vari : public vari {
+    /* dimension of the global parameters. */
+    int phi_size_;
+    /* global parameters. */
+    vari** phi_;
+    /* dimension of the latent Gaussian parameters. */
+    int theta_size_;
+    /* the marginal density of the observation, conditional on the 
+     * globl parameters. */
+    vari** marginal_density_;
+    /* An object to store the sensitivities of phi. */
+    Eigen::VectorXd phi_adj_;
+
+    template <typename T, typename K, typename D>
+    laplace_marginal_density_vari
+      (double marginal_density,
+       const Eigen::Matrix<T, Eigen::Dynamic, 1>& phi,
+       const std::vector<Eigen::VectorXd>& x,
+       const D& diff_likelihood,
+       const K& covariance_function,
+       const Eigen::VectorXd& theta,
+       const Eigen::VectorXd& W_root,
+       const Eigen::MatrixXd& L,
+       const Eigen::VectorXd& a,
+       const Eigen::VectorXd& l_grad)
+      : vari(marginal_density),
+        phi_size_(phi.size()),
+        phi_(ChainableStack::instance().memalloc_.alloc_array<vari*>(
+            phi_size_)),
+        theta_size_(theta.size()),
+        marginal_density_(
+          ChainableStack::instance().memalloc_.alloc_array<vari*>(1)) {
+      for (int i = 0; i < phi_size_; i++) phi_[i] = phi(i).vi_;
+
+      // CHECK -- is there a cleaner way of doing this?
+      marginal_density_[0] = this;
+      marginal_density_[0] = new vari(marginal_density, false);
```

Comment:
Not sure what's going on here but not sure this makes a lot of sense. `this` is a class while you setup enough memory for a `vari`.

Reply:
That we need to sit down and discuss properly.